modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/TransactionsInMultipleThreadsFailoverTest.java

```
@@ -0,0 +1,201 @@
+import org.apache.ignite.lang.IgniteClosure;
+import org.apache.ignite.testframework.GridTestUtils;
+import org.apache.ignite.transactions.Transaction;
+import org.apache.ignite.transactions.TransactionConcurrency;
+import org.apache.ignite.transactions.TransactionRollbackException;
+import org.apache.ignite.transactions.TransactionState;
+import org.junit.Assert;
+
+/**
+ *
+ */
+public class TransactionsInMultipleThreadsFailoverTest extends AbstractTransactionsInMultipleThreadsTest {
+    /**
+     * Starts transaction, breaks node and then resuming it in another thread.
+     *
+     * @param key Key to put.
+     * @param breakNodeIdx Node id to brake.
+     * @param initiatingNodeIdx Node, starting transaction on.
+     * @throws IgniteCheckedException If failed.
+     */
+    private void performTransactionFailover(String key,
+        int breakNodeIdx, int initiatingNodeIdx) throws IgniteCheckedException {
+
+        IgniteTransactions ts = grid(initiatingNodeIdx).transactions();
+
+        IgniteCache<String, Integer> cache = jcache(initiatingNodeIdx);
+
+        Transaction tx = ts.txStart(transactionConcurrency, transactionIsolation);
+
+        cache.put(key, 1);
+
+        tx.suspend();
+
+        G.stop(ignite(breakNodeIdx).name(), true);
+
+        Assert.assertNull(ts.tx());
+
+        IgniteInternalFuture<Boolean> fut = GridTestUtils.runAsync(new Callable<Boolean>() {
+            @Override public Boolean call() throws Exception {
+                Assert.assertNull(ts.tx());
+                Assert.assertEquals(TransactionState.SUSPENDED, tx.state());
+
+                tx.resume();
+
+                Assert.assertEquals(TransactionState.ACTIVE, tx.state());
+
+                tx.commit();
+
+                return true;
+            }
+        });
+
+        fut.get();
+    }
+
+    /**
+     * Starts tx locally with remote residing keys and then remote node fails.
+     */
+    public void testTxRemoteNodeFailover(){
```

let's throw an Exception if failed?